### ISSUE TITLE
[devel] Anonymous torrents not initializing in GUI

### ISSUE THREAD

**Issue by qstokkink** [Tribler/tribler#2581]
##### Tribler version/branch+revision: #####
3aa6bb204bbcc0cf7be3d6e8a355baf544ef6cb6
Since:
c4729596703f07d5512b4879605926056c22f666

##### Behavior: #####
I can no longer start anonymous downloads using the latest devel (3aa6bb204bbcc0cf7be3d6e8a355baf544ef6cb6): I can confirm they still work in 6.5.2 and up until 501b09f55479f11aacb6451d8064442895752ff4 (Oct/15/2016).
Downloads get stuck on "Initializing" and enabling/disabling multichain does not change the behavior.

A.F.A.I.K. there **should** not have been any changes breaking compatibility between the latest devel and 6.5.2 for the TunnelCommunity. Can anyone confirm this behavior?

**Comment by themusicgod1** [comment]
just grabbed c472959, as well as one of the later commits: issue exhibited itself.
Grabbed  501b09f : issue did not exhibit itself.  You appear to be correct about c472959 being the commit where this was introduced.  Looking at the commit it seems like it's missing a case for that particular state.

**Comment by qstokkink** [comment]
Thanks for confirming @themusicgod1. I did some digging in the commit log and I found:
```python
wantpeers.extend(self.guiUtility.library_manager.download_state_callback(no_collected_list))
```
was removed. I'm guessing this was removed in the process of separating the core from the GUI in preparation of the new GUI. However, this makes development on the `devel` branch impossible for anyone working with downloads.

@devos50 can we get a temporary fix (until the new GUI arrives) on the `devel` branch for this, so that people are still able to work with `devel`?
If you accept, the fix can be handled as follows. Insert the following on https://github.com/Tribler/tribler/blob/devel/Tribler/Core/APIImplementation/LaunchManyCore.py#L511:
```python
from Tribler.Main.vwxGUI.GuiUtility import GUIUtility
guiUtility = GUIUtility.getInstance()
if guiUtility.registered:
    guiUtility.library_manager.download_state_callback(states_list)
```

**Comment by devos50** [comment]
Hi guys,

Thanks a lot for debugging! Apparently, I overlooked the call to `self.guiUtility.library_manager.download_state_callback(no_collected_list)` and I only focussed on `wantpeers.extend` which I though was not very relevant. The latter method call indicates that libtorrent should return peer information when requesting info about a download.

I guess that the `download_state_callback` is doing some stuff that is necessary for us to correctly start a download. I will take a look tomorrow.

**Comment by qstokkink** [comment]
@devos50 Well, the only thing `guiUtility.library_manager.download_state_callback` does is update the GUI. I believe the download is still progressing in the background.

However, by not calling it, the old GUI is now effectively worthless: this makes debugging a lot harder. This is why I propose to leave the call in, until the new GUI is merged into devel.

**Comment by devos50** [comment]
@qstokkink if the download is still progressing in the background, it's only the GUI that is not refreshing fortunately. I don't want to create a temporary workaround by hard-linking the core package with the `vwxGUI` package so I propose to leave the state callback method in the GUI code. This way, the anon downloads should work again (in the GUI) and we don't make the core dependent on the GUI package.

I will place the code back in a PR I will try to merge today. If you have any better ideas, please let me know!

**Comment by qstokkink** [comment]
@devos50 The way I see it, you either put in place (1) a temporary fix in the core linking to the GUI or (2) a temporary fix* in the GUI to interface with the core. Tomayto, tomahto.

The reason I suggested option (1) is because it seems like it is part of the roadmap to the new GUI (read: smaller changelog in the PR).

Either way, something will have to be refactored before the new GUI gets merged, so I guess there is no best solution in this case.

<br>
\* I am assuming the new GUI uses a different method of grabbing the download states from the Core**
** Should the download state change perhaps be propagated through notifications?

**Comment by devos50** [comment]
@qstokkink having the 'fix' in the GUI makes it easier to just remove the old wx GUI as soon as we're ready to do that. I've implemented it in #2583 and tested it before and after the change and for me it works: the anon downloads are not stuck on initializing anymore.

**Comment by devos50** [comment]
@qstokkink I've merged the PR with the fix. Please let me know whether it also works for you now. If not, feel free to reopen this issue and I will take another look.
